File: ruby/lib/steam/sockets/steam_socket.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'ipaddr'
require 'socket'

require 'core_ext/stringio'
require 'exceptions/timeout_exception'

# This module implements common functionality for sockets used to connect to
# game and master servers
#
# @author Sebastian Staudt
module SteamSocket

  # The default socket timeout
  @@timeout = 1000

  # Sets the timeout for socket operations
  #
  # Any request that takes longer than this time will cause a
  # {TimeoutException}.
  #
  # @param [Fixnum] timeout The amount of milliseconds before a request times
  #        out
  def self.timeout=(timeout)
    @@timeout = timeout
  end

  # Creates a new UDP socket to communicate with the server on the given IP
  # address and port
  #
  # @param [String] ip_address Either the IP address or the DNS name of the
  #        server
  # @param [Fixnum] port The port the server is listening on
  def initialize(ip_address, port = 27015)
    @socket = UDPSocket.new
    @socket.connect ip_address, port
  end

  # Closes the underlying socket
  def close
    @socket.close
  end

  # Reads the given amount of data from the socket and wraps it into the buffer
  #
  # @param [Fixnum] buffer_length The data length to read from the socket
  # @raise [TimeoutException] if no packet is received on time
  # @return [Fixnum] The number of bytes that have been read from the socket
  # @see StringIO
  def receive_packet(buffer_length = 0)
    if select([@socket], nil, nil, @@timeout / 1000.0).nil?
      raise TimeoutException
    end

    if buffer_length == 0
      @buffer.rewind
    else
      @buffer = StringIO.alloc buffer_length
    end

    data = @socket.recv @buffer.remaining
    bytes_read = @buffer.write data
    @buffer.truncate bytes_read
    @buffer.rewind

    bytes_read
  end

  # Sends the given packet to the server
  #
  # This converts the packet into a byte stream first before writing it to the
  # socket.
  #
  # @param [SteamPacket] data_packet The packet to send to the server
  # @see SteamPacket#to_s
  def send(data_packet)
    puts "Sending data packet of type \"#{data_packet.class.to_s}\"." if $DEBUG

    @socket.send data_packet.to_s, 0
  end

end

File: php/lib/steam/sockets/SteamSocket.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2008-2011, Sebastian Staudt
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Sockets
 */

require_once STEAM_CONDENSER_PATH . 'ByteBuffer.php';
require_once STEAM_CONDENSER_PATH . 'UDPSocket.php';
require_once STEAM_CONDENSER_PATH . 'exceptions/TimeoutException.php';
require_once STEAM_CONDENSER_PATH . 'steam/packets/SteamPacketFactory.php';

/**
 * @package    Steam Condenser (PHP)
 * @subpackage Sockets
 */
abstract class SteamSocket
{
    /**
     * @var int
     */
    private static $timeout = 1000;

	/**
	 * @var ByteBuffer
	 */
	protected $buffer;

    /**
     * @var UDPSocket
     */
    protected $socket;

    /**
     * Sets the timeout for socket operations. This usually only affects
     * timeouts, i.e. when a server does not respond in time.
     *
     * Due to the server-side implementation of the RCON protocol, each RCON
     * request will also wait this amount of time after execution. So if you
     * need RCON requests to execute fast, you should set this to a adequatly
     * low value.
     *
     * @param $timeout The amount of milliseconds before a request times out
     */
    public static function setTimeout($timeout) {
        self::$timeout = $timeout;
    }

    public function __construct($ipAddress, $portNumber = 27015) {
        $this->socket = new UDPSocket();
        $this->socket->connect($ipAddress, $portNumber);
    }

    /**
     * Closes this socket
     *
     * @see #close
     */
    public function __destruct() {
        $this->close();
    }

    /**
     * Closes the underlying UDPSocket
     *
     * @see UDPSocket#close
     */
    public function close() {
        $this->socket->close();
    }

	/**
	 * Abstract getReplyData() method
	 * @return byte[]
	 */
	abstract public function getReply();

	/**
	 * @return int
	 */
	public function receivePacket($bufferLength = 0)
	{
        if(!$this->socket->select(self::$timeout)) {
			throw new TimeoutException();
		}

		if($bufferLength == 0)
		{
			$this->buffer->clear();
		}
		else
		{
			$this->buffer = ByteBuffer::allocate($bufferLength);
		}

        $data = $this->socket->recv($this->buffer->remaining());
        $this->buffer->put($data);
		$bytesRead = $this->buffer->position();
		$this->buffer->rewind();
		$this->buffer->limit($bytesRead);

		return $bytesRead;
	}

	/**
	 *
	 */
	public function send(SteamPacket $dataPacket)
	{
		trigger_error("Sending packet of type \"" . get_class($dataPacket) . "\"...");

        $this->socket->send($dataPacket->__toString());
	}
}
?>
